src/main/scala/org/constellation/rewards/RewardsManager.scala

```diff
@@ -1,11 +1,42 @@
-package org.constellation.util
+package org.constellation.rewards
 
+import cats.effect._
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import org.constellation.ConfigUtil
+import org.constellation.checkpoint.CheckpointService
+import org.constellation.consensus.Snapshot
+import org.constellation.domain.observation.Observation
 import org.constellation.p2p.PeerNotification
 import org.constellation.primitives.Schema.CheckpointEdge
 import org.constellation.primitives.{ChannelMessage, Transaction}
+import org.constellation.storage.{RecentSnapshot, SnapshotBroadcastService, SnapshotService}
 import org.constellation.trust.TrustEdge
 
-object Rewards {
+class RewardsManager[F[_]: Concurrent](
+  eigenTrust: EigenTrust[F],
+  snapshotService: SnapshotService[F],
+  snapshotBroadcastService: SnapshotBroadcastService[F],
+  checkpointService: CheckpointService[F]
+) {
+  final val rewards: Ref[F, Map[String, Double]] = Ref.unsafe(Map.empty)
+
+  def updateRewards(snapshot: Snapshot): F[Unit] =
+    for {
+      observations <- snapshot.checkpointBlocks.toList
+        .traverse(checkpointService.fullData)
+        .map(_.flatten.flatMap(_.checkpointBlock.observations))
+
+      _ <- eigenTrust.retrain(observations)
+      addressTrust <- eigenTrust.getTrustForAddresses
+
+      addresses = addressTrust.keySet.toList
+      distribution = RewardsManager.rewardDistribution(addresses, addressTrust)
+      _ <- rewards.modify(_ => (distribution, distribution))
```

Comment:
So there's a couple things to add/modify in updateRewards. First, at the end, we'll want to make an update to AddressService to increase the balance of the addresses of each node. Second, lets use rewards keep track of all historical rewards. When changing rewards, you could merge the maps and sum the values like 

`rewards.modify(prevDistro => (prevDistro ++ distribution.map{ case (k,v) => k -> (v + map1.getOrElse(k,0L)) }, ())`

note we'll probably want to change rewards to `Ref[F, Map[String, Long]]`.

Reply:
Fair point. Actually for historical rewards we can just use the `Semigroup` for `Map`:

```scala
rewards.modify(prevDistro => (prevDistro |+| newDistro, ()))
```

Reply:
About the `Double` to `Long` conversion we need to think about proper rounding so either `floor` or `ceil` or `round`.

Reply:
About the `Double -> `Long` we need to decide what to do with floating point. Any suggestions? :)